Reactor netty server adapter

crankydillo opened the pull request.

Commit:
Lots of hacks, but hello world 'does' work.

Commit:
Hacky request body reading.

Commit:
stuff

Commit:
gatling brings in netty:(

Commit:
Files were diffing because the byte[] is reused (duh).

Commit:
Return a Mono driven by the write-triggered subscription..

Will have to figure out what to do when there is no response.

Commit:
'deal' with body-less requests/responses.

Commit:
Really support completablefuture from asyncWrite..

related concept is solving the wait-for-subscription problem (what about
a timeout)?

Also add some support for writing headers.

Commit:
Just some initial messing around with timeout.  Setting this aside.

Commit:
More error handling.

Commit:
Header improvements when sending errors

Commit:
cleanup

Commit:
Start cleaning up adhoc test stuff..

Maybe the RestEasy folks will be ok with these put into test scope.

Commit:
This probably won't fly, but add a tomcat-based reference app...

that can be used to do adhoc comparison to the server-adapters.

Commit:
Cleanup, block on sync write.

Commit:
Add javadoc and rudimentary readme..

The readme assumes resteasy will take all this test-commons + adhoc
server stuff.  They may not.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+
+/**
+ * A server adapter built on top of <a
+ * href='https://github.com/reactor/reactor-netty'>reactor-netty</a>.  Similar
```
Since you provided a lnk to the docs for reactor, you could do the same here.

crankydillo replied:
Ah, I'd like to leave these as-is.  For some reason, I view the links I gave as the starting point.  Even though, they aren't consistent with one another.  If you feel strongly, I can change it.  That said, who looks at Javadoc anymore? :)

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+ * to the adapter built on top of netty4, this adapter will ultimately run on
+ * Netty rails.  Leveraging reactor-netty brings 2 main benefits, which are:
+ *
+ * 1. Reactor-netty's HttpServer + handle(req, resp) API is a little closer
```
"Reactor-netty" looks odd to me, I'd suggest one of the following:

- Reactor Netty
- reactor-netty

crankydillo replied:
I always struggle with this.  I will use the former!

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+ * 2. Reactor-netty puts a <a href='https://projectreactor.io/'>reactor</a>
+ * facade on top of Netty.  The Observable+Iterable programming paradigm is
+ * more general purpose than Netty's IO-centric Channel concept.  In theory, it
+ * should be more beneficial to learn:)
```
A third item could be to share resources between server and client side.

crankydillo replied:
I will list that, but I was really listing the benefits of this implementation given that a netty4 adapter already exists.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+ */
+public class ReactorNettyJaxrsServer implements EmbeddedJaxrsServer<ReactorNettyJaxrsServer> {
+
+   final Logger log = LoggerFactory.getLogger(ReactorNettyJaxrsServer.class);
```
Shouyldn't this be `private static final`?

crankydillo replied:
We need to capture all these comments, but just a heads up that I didn't concentrate on finer details like this.  I will give this another once-over after we get all the fundamentals correct.  However, if you want to point all of them out now, feel free:)  It will save me time later.

crankydillo replied:
On a related note.  I have a lot of TRACE calls in codepaths that get executed frequently.  I'm leaning towards removing all of those at the end due to be overly concerned about performance.  Do you have any concerns if those are left?

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+   final Logger log = LoggerFactory.getLogger(ReactorNettyJaxrsServer.class);
+
+   protected String hostname = null;
+   protected int configuredPort = PortProvider.getPort();
```
`PortProvider.getPort()` static call here may make testing difficult?

crankydillo replied:
It is controllable via a property.  I copied this from the netty adapter I think.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+   protected String root = "";
+   protected SecurityDomain domain;
+   private SSLContext sslContext;
+   private EmbeddedServerHelper serverHelper = new EmbeddedServerHelper();
```
`final`?

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+      return this;
+   }
+
+   public int getPort() {
```
There is no `@Override` on this one.  If not necessary, let's remove.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+   }
+
+   public int getPort() {
+      return runtimePort > 0 ? runtimePort : configuredPort;
```
Why would a runtime port be different than the configured one?  It will be helpful to add some documentation about these.  I would expect just one port variable.

crankydillo replied:
This stuff was just copied.  We can do a deeper dive on it.

crankydillo replied:
I would like to leave this.  I can see myself wanting to know what port was being used, especially if it came from config.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+   protected ResteasyDeployment deployment;
+   protected String root = "";
+   protected SecurityDomain domain;
+   private SSLContext sslContext;
```
This is never being used.  Does this mean the current setup only allows HTTP and not HTTPS?

Also, I do not see a setter in the interface.  In that case, can we consider a more immutable style here?  In general, I see the class is very mutable style, but that's mainly for limitation of existing code and interface..  but, we could consider minimize that..  

This is just a desire - and not waste time..  but, I would love to see a more builder style here.  Basically, the `start()` will be the `build()` class.  Internally, any `setPort`, or `setDeployment` would just build the builder.   I know the interface might be limiting it, but I guess we could mange it at implementation level..  I wish interface was not designed this way..

crankydillo replied:
I didn't concentrate at all on niceties like builder, not have I looked into HTTPS.  I'm hoping all that is just 'going to work'.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+      server = svrBuilder
+          .handle(handler::handle)
+          .bindNow();
+      runtimePort = server.port();
```
Hmmm..  port is dynamically picked?  We should be passing the port.

crankydillo replied:
The port is configured.  See line 68 above.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+          .bindNow();
+      runtimePort = server.port();
+
+      server.onDispose().block();
```
I have two questions here:

1. what does `onDispose()` itself do?  The Javadoc tells me that it just gives a Mono to tell us whether it completed or not.  I mean how do we use it?  It looks like it is not providing anything as of now?
2. `block()` will cause to block this thread, right?  Is that to prevent main thread existing?  Do we really need to do that?

crankydillo replied:
I was just doing this early on.  You are right that I need to change this.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+
+   class Handler {
+
+      private final Mono<InputStream> empty = Mono.just(new InputStream() {
```
This could potentially be `static` because we would need only one of it?  Or each subscription be done to a new instance?

crankydillo replied:
Only 1 `Handler` is instantiated for the 'server' so this is effectively a singleton (per server).

crankydillo replied:
Let me know if you really want this to be static.  I don't use as many statics as many seem to do:)

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+         // https://stackoverflow.com/a/51801335/2071683 but requires a thread.  Isn't using a thread
+         // per request even if from the elastic pool a big problem???  I mean we are trying to reduce
+         // threads!
+         // I honestly don't know what netty is doing here.  When I try to send a large body it says
```
let's quality "netty" in this comment as Netty JAX-RS Server.

crankydillo replied:
Will try to remember.  Also feel free to use the suggest feature:)  That said, I'm assuming all this will go away.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+        
+
+         // This is a subscription tied to the completion writing the response.
+         final MonoProcessor<Void> completionMono = MonoProcessor.create();
```
I am afraid of seeing Processor, cause I do not know much about those.

crankydillo replied:
Yeah, I mentioned this is only the second time I've had to deal with the 'observable' concept kind of 'spread out'. In this particular case, the MonoProcessor serves as both a subscription and a Mono.  The subscription is done way down where we actually get signals that the response is sent to the client.  Then the Mono part serves as what we return from `handle`.  It's kind of like chaining the subscriptions..  That said, I need to read the docs more closely on this.  I was going a bit fast in these areas.

 Still getting my mind around it.  If you can think of a better way, I'm all ears.  To be honest, I'm not entirely convinced cancellations and whatnot are working like I want.  I _think_ they are though.  I do have a backpressure problem, but that is not here.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+                     new ReactorNettyHttpResponse(req.method().name(), resp, completionMono);
+
+                 final ReactorNettyHttpRequest resteasyReq =
+                     new ReactorNettyHttpRequest(info, req, body, resteasyResp, (SynchronousDispatcher) deployment.getDispatcher());
```
I need to check what `SynchronousDispatcher` vs async.  I guess sync one just executes within the same thread instead of jumping to a different thread pool?

crankydillo replied:
I didn't think much about this.  The doc/javadoc for resteasy is non-existent (or at least I couldn't find it, and they didn't answer on the issue I wrote).  I basically just copied this from what I saw with the netty adapter.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+                 }
+                 return completionMono;
+             }).onErrorResume(t -> {
+                resp.status(500).sendString(Mono.just(t.getLocalizedMessage())).subscribe(completionMono);
```
Is this what we do for Tomcat as a response?  if so, totally fine.

But, we may want to add a comment that resuming is on purpose.

crankydillo replied:
I may have been premature.  The goal was to avoid server collapse upon an individual error, but I'm guessing reactor-netty takes care of that.  We can test this.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+                 // This is what actually kick RestEasy into action.
+                 deployment.getDispatcher().invoke(resteasyReq, resteasyResp);
+
+                 if (!resteasyReq.getAsyncContext().isSuspended()) {
```
hmm..  so, for now we only support the blocking scenarios or is this specifically for the `@Suspended`?

crankydillo replied:
I still need to look into this 'suspension' concept.  Anyhow, let's discuss this in person.  We may have significant issues here:(

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,209 @@
+
+                 if (!resteasyReq.getAsyncContext().isSuspended()) {
+                     try {
+                         resteasyResp.finish();
```
This suggests that we are only dealing with blocking scenarios for now?  Because the response should be ready to flush and close ?

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyHttpRequest.java:
```diff
@@ -0,0 +1,414 @@
+    @Override
+    public void setInputStream(InputStream stream) {
+        // TODO what is this about?
+        this.in = in;
```
did you mean `this.in= stream`?

anilgursel reviewed (commented):
Just reviewed `ReactorNettyJaxrsServer` so far.  I still need to go through request/response objects.

Commit:
Primarily logging additions.

Commit:
Endpoint improvements.

Commit:
Take 2 at truly supporting CompletableFuture (still sort of)...

on asyncWrite.  For some crazy reason, I thought a lambda function would
'reach' out to surrounding scope every time it was called.

Commit:
PR feedback

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyHttpRequest.java:
```diff
@@ -0,0 +1,417 @@
+            // check if this CF is already resolved
+            CompletableFuture<Void> ret = f.toCompletableFuture();
+            // if it's not resolved, we may need to suspend
+            if(!ret.isDone() && !isSuspended()) {
```
I do not have any specific suggestion..  But, seeing this (even if it is `volatile`) worries me about potential race conditions.

crankydillo replied:
This is all just copied from netty4.  We have a task about working through all this stuff.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyHttpRequest.java:
```diff
@@ -0,0 +1,417 @@
+
+        @Override
+        public CompletionStage<Void> executeBlockingIo(RunnableWithException f, boolean hasInterceptors) {
+            if(1 == 1) { // TODO if(!NettyUtil.isIoThread()) {
```
What's the reason to check specifically for IO thread?   There may be other threads that we may not want to block.

Does it expect to move to a different thread pool by this time?

crankydillo replied:
As with a lot of this code, it is just copied from netty and needs to be reviewed.  Before I rolled off of this, I was trying to understand when this method would be called.  I found the conditional in resteasy code, but I wasn't sure how to get into the state such that this would be called.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyHttpRequest.java:
```diff
@@ -0,0 +1,417 @@
+                    return ret;
+                }
+                return CompletableFuture.completedFuture(null);
+            } else if(!hasInterceptors) {
```
What's the reason behind this check?  If there are interceptors, we can't use it on ioThread?

crankydillo replied:
Last one of these I'll respond to.  Have to go through this code:)

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyHttpRequest.java:
```diff
@@ -0,0 +1,417 @@
+                if(!isSuspended()) {
+                    suspend();
+                }
+                return CompletableFuture.runAsync(() -> {
```
Should this really run on `ForkJoinPool#commonPool()`?  This is some blocking code, so I'd expect this to run on some thread pool which is targeted for blocking calls.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyHttpRequest.java:
```diff
@@ -0,0 +1,417 @@
+
+            @Override
+            public void complete() {
+                synchronized (responseLock)
```
This is `synchronized` because we expected `complete` to be called from multiple locations at the same time?  If so, is there a happy path scenario for it, or is it to cover some exception/error scenarios?  In any case, if this `synchoronized` is really synchronizing some work, then that's a problem as it would impact the efficiency.

I guess we need to better understand where it comes as useful and question whether really needed or not.

I have the same comment for resume/cancel.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java:
```diff
@@ -0,0 +1,186 @@
+   private final HttpServerResponse response;
+
+   /**
+    * A signal back to the 'main' Flux that writing bytes back to the client has finished.
```
Would you mind explaining in the comments what `'main'` refers to here?

crankydillo replied:
Sure thing.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java:
```diff
@@ -0,0 +1,186 @@
+   /**
+    * Indicates we've started sending bytes.
+    */
+   private boolean started; // want to eliminate this
```
Can you please add a `// TODO` here?

crankydillo replied:
I will remove it.  I don't know how to eliminate it right now.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java:
```diff
@@ -0,0 +1,186 @@
+   @Override
+   public CompletionStage<Void> asyncFlush() {
+      // TODO
+      return CompletableFuture.completedFuture(null);
```
I guess this would cause some problems because flush is not working for now?

crankydillo replied:
I've been wondering what 'flush' means in in the reactor world.  I still don't know.  It remains a TODO, but hopefully I can get to it within a few days.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java:
```diff
@@ -0,0 +1,186 @@
+
+    @Override
+   public CompletableFuture<Void> asyncWrite(final byte[] bs, int offset, int length) {
+        if (!started) {
```
I assume we would not call `asyncWrite` concurrently from different threads (talking about the same response).

crankydillo replied:
I don't think that will happen in the 'normal' case, where RestEasy doesn't make any thread changes.  We could read the code or ask reactor-netty folks about that.  On RestEasy thread switches, I don't know about how all the thread paths work.  Later on you ask several questions about the 'suspend'/'resume'/etc stuff.  Almost all of that was blindly copied from the netty4 adapter.  It remains a TODO to understand that.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java:
```diff
@@ -0,0 +1,186 @@
+                actualOut = actualOut.timeout(timeout);
+            }
+            response
+                .sendByteArray(actualOut)
```
This seems to be using `Unpooled::wrappedBuffer`.  Any specific reason not using `send(Publisher<? extends ByteBuf> dataStream)` API ?

crankydillo replied:
I just assumed that since we have `byte[]` that reactor-netty would do the best thing with `Flux<byte[]>`.  Seems weird that we'd do a better job converting it to a `ByteBuf`, but we can certainly look into it.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java:
```diff
@@ -0,0 +1,186 @@
+            .map(ignore -> {
+                byte[] bytes = bs;
+                if (offset != 0 || length != bs.length) {
+                    bytes = Arrays.copyOfRange(bs, offset, offset + length);
```
Byte copy gets expensive.  Is there anyway not do the copy?

crankydillo replied:
I gave it a brief think, but couldn't come up with anything.  It only gets copied in that conditional case.  Adhoc testing suggests that's only on the last call to `asyncWrite`.  I have to be honest though.  I did a lot of this dev when working with a returned `InputStream`.  I'm guessing in our normal case (serialized pojos), you'll get one big byte[] where that condition would not hold true.  We can test though.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java:
```diff
@@ -0,0 +1,186 @@
+                out.map(tuple -> {
+                    totalBytesSent += tuple.getT1().length;
+                    log.trace("{} total bytes sent", totalBytesSent);
+                    tuple.getT2().complete(COMPLETED_SIGNAL);
```
So, when we hit this line the, we would consider `asyncWrite` to be completed (I mean the future it returns is complete).  We would not hit this line if downstream did not request elements..  So, I guess this is as close as we can.  No concerns to point at this moment.

crankydillo replied:
Yes, and it is ultimately flatMapped over in the `Mono` we return.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java:
```diff
@@ -0,0 +1,186 @@
+   public CompletableFuture<Void> asyncWrite(final byte[] bs, int offset, int length) {
+        if (!started) {
+            started = true;
+            Flux<byte[]> actualOut =
```
So, this is a `Flux` because we expect `asyncWrite` to be called multiple time for a response?  I assume that's true for SSE scenarios, or potentially other streaming/chunking scenarios.

crankydillo replied:
I haven't tested SSE, but yes it happens for chunking scenarios.

anilgursel commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java:
```diff
@@ -0,0 +1,186 @@
+   /**
+    * The {@link Flux<byte[]>} that is fed into {@link HttpServerResponse#sendByteArray}.
+    */
+   private Flux<Tuple2<byte[], CompletableFuture<Boolean>>> out = Flux.create(sink -> {
```
Looks like this may lead to a OOM if there is one response that's really huge.  I am not sure if other `OverflowStrategy` options would work though..  except may be erroring..  Maybe we can keep `BUFFER`; however, then do a check somewhere on the size and give an error if it is beyond buffer size..

crankydillo replied:
You are worried about the case where this no downstream demand, right?  Anyhow, there is concern about OOM.  I believe the netty adapter restricts payload size.

anilgursel reviewed (commented):
Summarizing for making it for myself easier to remember the important items:

- [ ] I am concerned about `synchronized` in `ReactorNettyHttpRequest`.
- [ ] Suspending if not suspend may introduce race conditions.
- [ ] Quite a few `// TODO` items.
- [ ] `Unpooled::wrappedBuffer` usage in sendByteArray.

crankydillo reviewed (commented):
I didn't really answer all the suspend/resume/etc questions.  We have a task to look at that more closely.  I hope to start on it today or tomorrow, after setting up some rudimentary LnP.

Commit:
Address some PR comments.

Commit:
I thought I used a wildcard import:)

Commit:
Use @anilgursel 's EmitterProcessor.

I knew I was missing something when I had to create my own listener!

Anyhow, I can see no negatives to EmitterProcessor and the code is
significantly simpler because I don't have to worry about how the sink
and the listener have to be tied together.

Commit:
Copied (yuck!) the tests from the netty4 adapter.  All are passing...

except the server side idle timeout one.  This actually has me worried a
bit, but it may just be me not understanding how to modify the channel
pipeline correctly.

Commit:
So intellij does not do perfect refactoring!

Commit:
All tests are passing.

lalifk commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,319 @@
+              .port(configuredPort);
+
+      if (hostname != null && !hostname.trim().isEmpty()) {
+         svrBuilder.host(hostname);
```
Is it safer to assign `svrBuilder` back to itself? e.g.`svrBuilder = svrBuilder.host(hostname)`

crankydillo replied:
@lalifk First, thanks for reviewing!  Second, sorry this is so delayed.  I actually went on a long vacation right after submitting this.

This work has moved to [this branch](https://github.com/crankydillo/Resteasy/tree/RESTEASY-2700_reactor-netty-server-adapter) where we are nearing to making a PR to the official repo.

lalifk commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,319 @@
+
+      Publisher<Void> handle(final HttpServerRequest req, final HttpServerResponse resp) {
+
+         if (1 == 2) {
+            final NettyOutbound o =
+                resp.send(req.receive().retain().doOnEach(b -> log.trace("got some bytes!")));
+            return o.then()
+                .doFinally(s -> log.trace("Request processing finished with: {}", s));
+         }
```
Temporary code I suppose?

lalifk commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,319 @@
+             .flatMap(body -> {
+                log.trace("Body read!");
+                resp.withConnection(conn -> {
+                   log.trace("{}", conn.channel().id());
```
Better add a message before to describe the value that's being logged

lalifk commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,319 @@
+             .aggregate()
+             .asInputStream()
+             .switchIfEmpty(empty)
+             .flatMap(body -> {
```
[Minor] The body of this lambda expression is quite long. I foresee it being flagged by sonar...

crankydillo replied:
Unsurprisingly, it has grown.  I may isolate this function in the new branch.  Thanks!

lalifk commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,319 @@
+                   ThreadLocalResteasyProviderFactory.push(deployment.getProviderFactory());
+                }
+
+                deployment.getDispatcher().invoke(resteasyReq, resteasyResp);
```
Please help me understand: At what point is `deployment` initialized? Is it through earlier invocations of `getDeployment()` by some other class?

crankydillo replied:
[Code](https://github.com/crankydillo/Resteasy/blob/2340a5467038d517c04f5efcfae6131040915aed/server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java#L250)

lalifk commented on server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ReactorNettyJaxrsServer.java:
```diff
@@ -0,0 +1,319 @@
+
+                deployment.getDispatcher().invoke(resteasyReq, resteasyResp);
+
+                if (defaultInstance instanceof ThreadLocalResteasyProviderFactory) {
```
Somewhere before this line, we should be setting the `sslContext` it seems. I think follow up PRs will address it

crankydillo replied:
Yes!  @sharathsrinivasa just contributed [that](https://github.com/crankydillo/Resteasy/pull/2).

crankydillo commented:
This work was moved to [this branch](https://github.com/crankydillo/Resteasy/tree/RESTEASY-2700_reactor-netty-server-adapter) as we prepare to make a PR to the official repo.
